File: Kinematics/python/interface/interface.py
import socket
import numpy as np
import math
class Interface:
    class Data:
        pass
    def __init__(self):
        self.recv_port = 45454
        self.send_port = 45455
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.socket2 = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        self.socket2.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        self.socket.bind(('', self.recv_port))
        self.cmd = self.Data()
    
    def getSignedCharValue(self,a):
        a = a%256
        if a > 127:
            return a - 256
        return a

    def getVelCmd(self):
        data, _ = self.socket.recvfrom(1024)
        self.cmd.velX = self.getSignedCharValue(data[0])
        self.cmd.velY = -self.getSignedCharValue(data[1])
        self.cmd.velW = self.getSignedCharValue(data[2])
        return self.cmd
        
    def controlWheel(self,a,b,c,d):
        data = np.array([a,b,c,d])*60/(2*math.pi)/2.5
        # data = np.array([a,b,c,d])
        data = data.clip(-127,127)
        data = bytes(data.astype(np.int8))
        self.socket2.sendto(data,('127.0.0.1',self.send_port))

File: Kinematics/matlab/Interface.m
classdef Interface
    properties
        recv_port = 45454
        send_port = 45455
        socket
    end
    methods
        function obj = Interface()
            % there's only one interface in the initailization stage.
            obj.socket = udpport("datagram","IPV4","LocalPort", obj.recv_port);
        end
        function data = getVelCmd(obj)
            while true
                if (obj.socket.NumDatagramsAvailable > 0)
                    datagram = read(obj.socket, obj.socket.NumDatagramsAvailable, "int8");
                    data = struct();
                    datagram;
                    data.velX = datagram(1).Data(1);
                    data.velY = -datagram(1).Data(2);
                    data.velW = datagram(1).Data(3);
                    return
                end
            end
        end
        function controlWheel(obj,v1,v2,v3,v4)
            % TODO convert from floating to uint8
            d = [v1,v2,v3,v4]*60/(2*math.pi)/2.5;
            d(d>127) = 127;
            d(d<-127) = -127;
            data = int8(d);
            data
            write(obj.socket,data,"int8","127.0.0.1",obj.send_port);
        end
        function delete(obj)
            clear obj.socket;
        end
    end
end
